src/app/rosetta/lib/mempool.ml

```diff
@@ -131,46 +142,174 @@ module Transaction = struct
               graphql_uri )
       ; validate_network_choice= Network.Validate_choice.Real.validate }
 
+    let obj_of_user_command_info (user_command_info : User_command_info.t) =
+      object
+        (* TODO: Swap Id with hash *)
+        method id = user_command_info.hash
+
+        method amount =
+          Option.value ~default:Unsigned.UInt64.zero user_command_info.amount
+
+        method fee = user_command_info.fee
+
+        method kind =
+          match user_command_info.kind with
+          | `Payment ->
+              `String "PAYMENT"
+          | `Delegation ->
+              `String "STAKE_DELEGATION"
+          | `Create_token ->
+              `String "CREATE_NEW_TOKEN"
+          | `Create_account ->
+              `String "CREATE_TOKEN_ACCOUNT"
+          | `Mint_tokens ->
+              `String "MINT_TOKENS"
+
+        method feeToken = user_command_info.fee_token
+
+        method feePayer =
+          object
+            method publicKey =
+              let (`Pk p) = user_command_info.fee_payer in
+              `String p
+          end
+
+        method nonce = Unsigned.UInt32.to_int user_command_info.nonce
+
+        method receiver =
+          object
+            method publicKey =
+              let (`Pk p) = user_command_info.receiver in
+              `String p
+          end
+
+        method source =
+          object
+            method publicKey =
+              let (`Pk p) = user_command_info.source in
+              `String p
+          end
+
+        method token = user_command_info.token
+      end
+
     let mock : 'gql Mock.t =
       { gql=
           (fun ~public_key:_ ->
-            (* TODO: Add variants to cover every branch *)
             Result.return
             @@ object
                  method pooledUserCommands =
-                   [| `UserCommand
-                        (object
-                           method id = "TXN_1"
-                        end)
-                    ; `UserCommand
-                        (object
-                           method id = "TXN_2"
-                        end) |]
+                   User_command_info.dummies
+                   |> List.map ~f:(fun info ->
+                          `UserCommand (obj_of_user_command_info info) )
+                   |> List.to_array
                end )
       ; validate_network_choice= Network.Validate_choice.Mock.succeed }
   end
 
   module Impl (M : Monad_fail.S) = struct
-    module E = Env.T (M)
+    let user_command_info_of_obj obj =
+      let open M.Let_syntax in
+      let extract_public_key data =
+        match data#publicKey with
+        | `String pk ->
+            M.return (`Pk pk)
+        | x ->
+            M.fail
+              (Errors.create
+                 ~context:
+                   (sprintf
+                      "Received a public key of an unexpected shape %s when \
+                       accessing the internal Coda GraphQL API."
```

Comment:
Nit: I would remove 'internal' here, in favour of `Coda daemon's` or similar. The implication of internal is that it is not an API for public use: I think this is not true.